### PR TITLE
[5/n][FSDP] Update _sharded_post_state_dict_hook to use DTensor when use_dtensor=True in state_dict_config

### PR DESCRIPTION
**Pull request by wz337**
This allows us use use_dtensor=True for ShardedStateDictConfig() before calling model.state_dict().

load_state_dict hooks updates will be in next PR. 

cc @zhaojuanmao @mrshenli @rohan-varma @awgu

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_dtensor_state_dict.py [hunk 1]
```diff
@@ -87,5 +86,49 @@ def test_dtensor_sharded_optim_state_dict(self):
...
+
+        for dtensor_sd, sharded_tensor_sd in zip(
+            dtensor_sd.items(), sharded_tensor_sd.items()
+        ):
+            k1, v1 = dtensor_sd
+            k2, v2 = sharded_tensor_sd
+            if isinstance(v1, DTensor) and isinstance(v2, ShardedTensor):
+                self.assertEqual(k1, k2)
+                # check whether local_tensor are the same
+                self.assertEqual(v1.to_local(), v2.local_shards()[0].tensor)
```
Comment (fduwjj) [inline test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:126]: you can use .local_tensor() of ShardedTensor to get `local_shards()[0].tensor`

**Review by fduwjj — APPROVED** [review]
LGTM

File: torch/distributed/fsdp/_shard_utils.py [hunk 2]
```diff
@@ -142,10 +142,10 @@ def _create_chunk_dtensor(
...
     shard_placement = DShard(0)
-    tensor_padded_list, _ = shard_placement._split_tensor(
+    tensor_list, _ = shard_placement._split_tensor(
         tensor,
         device_mesh.size(dim=0),
         with_padding=False,
         contiguous=True,
     )
-    return DTensor.from_local(tensor_padded_list[rank], device_mesh, [shard_placement])
+    return DTensor.from_local(tensor_list[rank].clone(), device_mesh, [shard_placement])
```
Comment (fegin) [inline torch/distributed/fsdp/_shard_utils.py:151]: Let's add a comment why clone is required. Also, not sure if we need to call `detach()`. State dict hooks may be called inside `no_grad` context. But it is probably safer to call `detach()`.

**Review by fegin — APPROVED** [review]
LGTM
